lib/tapioca/gem/events.rb

```diff
@@ -42,6 +42,19 @@ def initialize(symbol, constant)
       end
     end
 
+    class NotOwnedConstantFound < ConstantFound
+      extend T::Sig
+
+      sig { returns(Module).checked(:never) }
+      attr_reader :constant
+
+      sig { params(symbol: String, constant: Module).void.checked(:never) }
+      def initialize(symbol, constant)
+        super
+        @constant = constant
```

Comment:
I wonder if we should do the breakdown of the kinds of constants at push time, so that these become 4 separate events instead:
1. `AliasConstantFound`
2. `ValueConstantFound`
3. `ModuleConstantFound`
4. `ForeignModuleConstantFound < ModuleConstantFound`

All we would need to do is to move the logic in `compile_constant` inside `push_constant` and the handling becomes much simpler. WDYT @Morriar?

Reply:
Yeah I like it!

Another option to keep the original design is to use a cast:

```rb
class ForeignConstantFound < ConstantFound
  extend T::Sig

  sig { override.returns(Module) }
  def constant
    T.cast(@constant, Module)
  end

  sig { params(symbol: String, constant: Module).void }
  def initialize(symbol, constant)
    super
  end
end
```

I'm good with both approaches 👍

Reply:
I created an issue to follow up on this: #977